### PR TITLE
[v14] refactor: example apps to use hooks where possible

### PR DESCRIPTION
**Pull request by emmenko**

### CODE REVIEW

File: playground/src/components/state-machines-details/state-machines-details.js [hunk 1]
```diff
@@ -15,84 +15,116 @@ import {
...
+  isLoading: true,
+  error: null,
+  data: null,
+};
+
+// NOTE: using `useEffect` for asynchronously update the state, thus re-rendering
+// the component, has currently a limitation when testing the component, as it throws
+// a warning about `act`. This will be fixed in react@16.9.
+// In the meantime we can use a plain old stateful class component.
+class StateMachinesDetailsFetcher extends React.Component {
```
Comment (tdeekens) [inline playground/src/components/state-machines-details/state-machines-details.js]: Why did this compnonent change at all?

File: playground/src/components/state-machines-details/state-machines-details.js [hunk 2]
```diff
@@ -15,84 +15,116 @@ import {
...
+const StateMachinesDetails = props => {
+  const dataLocale = useApplicationContext(context => context.dataLocale);
+  const dispatch = useDispatch();
+  const showApiErrorNotification = useShowApiErrorNotification();
+
+  return (
+    <StateMachinesDetailsFetcher
+      id={props.id}
+      dataLocale={dataLocale}
+      dispatch={dispatch}
```
Comment (tdeekens) [inline playground/src/components/state-machines-details/state-machines-details.js]: Ah, can we not separate concerns here and make the child less aware and not pass `dispatch` but a bound action creator already?

File: playground/src/components/state-machines-list/state-machines-list.js [hunk 3]
```diff
@@ -35,107 +34,84 @@ export const columnsDefinition = [
...
-  measurementCache = null;
-  registerMeasurementCache = cache => {
-    this.measurementCache = cache;
-  };
-  handleRowClick = (rowIndex, results) => {
-    this.props.goToStateMachineDetail(results[rowIndex].id);
-  };
+const StateMachinesList = props => {
+  const cachedStateMachine = useSelector(selectStateMachinesFromCache);
+  const cachedStateMachineObjectsCount = cachedStateMachine
```
Comment (tdeekens) [inline playground/src/components/state-machines-list/state-machines-list.js]: Isn't this just `hasStateMachines` as a better name?

File: playground/src/components/state-machines-list/state-machines-list.spec.js [hunk 4]
```diff
@@ -81,11 +98,87 @@ describe('list view', () => {
...
+                },
+              },
+            },
+            response: {
+              ...defaultMock.response,
+              id: 'sm2',
+              key: 'sm-2',
+            },
+          },
+        ],
```
Comment (tdeekens) [inline playground/src/components/state-machines-list/state-machines-list.spec.js]: Maybe we can keep the setup here a bit slimmer in that we have different mocks define above?

File: application-templates/starter/src/routes.js [hunk 5]
```diff
@@ -1,28 +1,38 @@
...
 
-const ApplicationRoutes = ({ match }) => (
-  <RestrictedByPermissions
-    permissions={[PERMISSIONS.ManageProducts, PERMISSIONS.ViewProducts]}
-    unauthorizedComponent={PageUnauthorized}
-    shouldMatchSomePermissions={true}
-  >
+const ApplicationRoutes = ({ match }) => {
+  const canRenderRestrictedRoutes = useIsAuthorized({
+    demandedPermissions: [PERMISSIONS.ManageProducts, PERMISSIONS.ViewProducts],
```
[the remark this answers is not in the document]
Comment (emmenko) [reply application-templates/starter/src/routes.js]: You're right, I'll fix that thanks

File: playground/src/components/state-machines-details/state-machines-details.js (hunk 1 above)
Comment (emmenko) [reply playground/src/components/state-machines-details/state-machines-details.js]: Yeah so actually I wanted to use only hook (with `useEffect`) but there is still this issue with `act` in testing. So I just extracted the fetching logic into a stateful component, but it's just a temporary thing. Eventually it should all be hooks.

File: playground/src/components/state-machines-details/state-machines-details.js (hunk 2 above)
Comment (emmenko) [reply playground/src/components/state-machines-details/state-machines-details.js]: See https://github.com/commercetools/merchant-center-application-kit/pull/856#discussion_r303628451

File: playground/src/components/state-machines-list/state-machines-list.js (hunk 3 above)
Comment (emmenko) [reply playground/src/components/state-machines-list/state-machines-list.js]: Hmm what do you mean? The `cachedStateMachineObjectsCount` is a number.

File: playground/src/components/state-machines-details/state-machines-details.js (hunk 1 above)
Comment (tdeekens) [reply playground/src/components/state-machines-details/state-machines-details.js]: Yeah I know this problem. Haf it with sign up components too.

File: playground/src/components/state-machines-details/state-machines-details.js (hunk 2 above)
Comment (tdeekens) [reply playground/src/components/state-machines-details/state-machines-details.js]: That part is clear. I was just wondering is we can not pass disptach down but bound action creators.

Comment (emmenko) [reply playground/src/components/state-machines-details/state-machines-details.js]: Changed, thanks

File: application-templates/starter/src/routes.js [hunk 6]
```diff
@@ -1,28 +1,38 @@
...
+);
 
-const ApplicationRoutes = ({ match }) => (
-  <RestrictedByPermissions
-    permissions={[PERMISSIONS.ManageProducts, PERMISSIONS.ViewProducts]}
-    unauthorizedComponent={PageUnauthorized}
-    shouldMatchSomePermissions={true}
-  >
+const ApplicationRoutes = ({ match }) => {
+  const canRenderRestrictedRoutes = useIsAuthorized({
```
Comment (tdeekens) [inline application-templates/starter/src/routes.js]:
```suggestion
  const vanViewProducts = useIsAuthorized({
```

File: playground/src/components/state-machines-details/state-machines-details.js [hunk 7]
```diff
@@ -1,8 +1,8 @@
 import React from 'react';
 import PropTypes from 'prop-types';
-import flowRight from 'lodash/flowRight';
-import { withApplicationContext } from '@commercetools-frontend/application-shell-connectors';
-import { connect } from 'react-redux';
+import { useDispatch } from 'react-redux';
+import { useApplicationContext } from '@commercetools-frontend/application-shell-connectors';
```
Comment (tdeekens) [inline playground/src/components/state-machines-details/state-machines-details.js]: Are these used in this component. Feels like I fail to spot that.

File: playground/src/components/state-machines-details/state-machines-details.js [hunk 8]
```diff
@@ -15,84 +15,120 @@ import {
...
-    }
-  )
-)(StateMachinesDetails);
+const StateMachinesDetails = props => {
+  const dataLocale = useApplicationContext(context => context.dataLocale);
+  const dispatch = useDispatch();
+  const fetchData = React.useCallback(
+    (...args) => dispatch(fetchStateMachine(...args)),
+    [dispatch]
+  );
```
Comment (tdeekens) [inline playground/src/components/state-machines-details/state-machines-details.js]: Thanks.

File: playground/src/components/state-machines-details/state-machines-details.js (hunk 7 above)
Comment (emmenko) [reply playground/src/components/state-machines-details/state-machines-details.js]: Yes, for the `dataLocale`

**Pull request merged**